### ISSUE TITLE
[RFC] Local QA Coprocessor — SLM <2B for Build/Test Triage

### ISSUE THREAD

**Issue by Luciansvon** [Luciansvon/B.I.M.A-DEV-INFRA#9]
# Local QA Coprocessor

> **Type:** RFC / RESEARCH-BACKED DESIGN  
> **Status:** PROPOSED / BLOCKED BY CONSUMER + BENCHMARK DATASET  
> **Domain:** AGENT-COST / CI-QA / LOCAL-SLM  
> **Primary candidate:** `Qwen2.5-Coder-1.5B Base`

---

## One-line concept

Tambahkan **SLM lokal <2B sebagai `CHEAP/SPECIALIZED AGENT`** di antara deterministic verification dan frontier agent untuk mengurangi pekerjaan reasoning mahal pada build/test/release QA.

Model **bukan sumber kebenaran** dan tidak boleh memberi verdict production-ready sendiri.

---

## Relationship to existing architecture

Issue ini merupakan implementasi lanjutan dari slot `CHEAP/SPECIALIZED AGENT` pada #3.

```text
MACHINE
  ↓
bima-evidence.v1
  ↓
bima-agent-packet.v1
  ↓
Local QA Coprocessor <2B
  ↓
confidence / policy gate
  ├─ high + bounded → structured result
  └─ low / unknown → STRONG AGENT
```

ReleaseProof #8 menjadi kandidat consumer pertama setelah fondasi shared infra stabil.

---

# Scope

Local QA Coprocessor boleh menangani task sempit seperti:

- build/test failure classification;
- primary error extraction;
- log summarization dari evidence yang sudah dipreprocess;
- dependency/compiler/environment/test failure categorization;
- bounded diff review;
- historical-fix retrieval/ranking;
- structured JSON output;
- escalation recommendation.

Model tidak boleh menjadi primary authority untuk:

- compiler/test truth;
- security scanning;
- artifact/provenance verification;
- architecture review;
- complex multi-file debugging;
- production/release verdict final.

---

# Model decision

## Primary foundation

`Qwen2.5-Coder-1.5B Base`

Alasan pemilihan:

- code-specialized foundation;
- ukuran masih sesuai target local SLM;
- tersedia Base untuk continued pretraining/SFT;
- evidence fine-tuning code + execution feedback lebih matang dibanding kandidat tiny general model;
- cocok dengan domain build logs, compiler errors, test failures, dependency/config issues, dan bounded code context.

## Required benchmark competitors

Sebelum model final dikunci, benchmark minimal:

1. `Qwen2.5-Coder-1.5B`
2. `Granite 4.0 1B Dense`
3. `Qwen3.5-0.8B Base`

`Qwen3.5-0.8B` diposisikan terutama sebagai target distillation / Nano variant setelah model utama terbukti.

Community distill berlabel Qwen3.8 tidak menjadi foundation default sebelum lolos benchmark internal dan provenance/training quality-nya cukup jelas.

---

# Required prerequisite

**Jangan implementasikan runtime/training SLM sebelum:**

- [x] `zizmor` increment dari #3 selesai;
- [x] `bima-evidence.v1` canonical evidence tersedia;
- [x] `bima-agent-packet.v1` tersedia;
- [ ] relevant log extraction/preprocessing deterministic tersedia;
- [ ] minimal satu real consumer terintegrasi;
- [ ] benchmark dataset punya held-out split yang tidak masuk training.

SLM harus menerima packet/evidence kecil, bukan raw repository + puluhan ribu baris log.

---

# ReleaseProofBench / LocalQABench

Sebelum training, buat benchmark internal minimal 1,000 kasus terverifikasi.

Target distribusi awal:

```text
300 build failures
200 test failures
150 dependency failures
100 environment/toolchain failures
100 packaging/release failures
100 noisy/misleading logs
100 healthy builds
100 novel/unknown failures
```

Dataset dapat tumbuh lintas project, tetapi project-specific data tetap dimiliki project asal. Shared benchmark hanya menyimpan schema, reusable patterns, atau data yang memang aman/promotable sesuai project contract.

---

# Metrics

Jangan menilai dengan kualitas tulisan.

Minimal ukur:

```text
classification_macro_f1
root_cause_top1
root_cause_top2
critical_false_negative_rate
hallucinated_evidence_rate
valid_json_rate
escalation_recall
confidence_calibration
latency
RAM / VRAM
frontier_escalation_rate
frontier_context_reduction
frontier_calls_saved
```

Critical false negative harus menjadi metric prioritas: failure nyata tidak boleh berubah menjadi false-green hanya karena SLM percaya diri.

---

# Training path

```text
Qwen2.5-Coder-1.5B Base
          ↓
continued pretraining (optional, measured)
          ↓
verified CI/build/test corpus
          ↓
SFT
          ↓
execution/verifier feedback
          ↓
ReleaseProof-QA-1.5B
          ↓
optional distillation
          ↓
Qwen3.5-0.8B student
```

Training examples idealnya mempunyai:

```text
INPUT
- normalized evidence
- stage
- environment/toolchain
- relevant diff/files
- test/build result

TARGET
- failure class
- evidence references
- root-cause candidate
- recommended next action
- escalate true/false

VERIFICATION
- actual fix
- rebuild/retest result
```

Compiler, tests, lint, artifact checks, dan policy engine tetap menjadi verifier utama.

---

# Failure Memory

Setelah frontier agent menyelesaikan failure yang belum dikenal dan verification PASS, simpan reusable pattern bila sesuai ownership rule:

```text
failure signature
→ evidence
→ confirmed root cause
→ fix class
→ verification result
```

Local SLM kemudian dapat membantu rank/retrieve historical cases sebelum eskalasi.

Project-specific incident detail tetap di project repository. Promote ke global hanya jika root cause shared infra atau berulang lintas >=2 project sesuai `AGENTS.md`.

---

# Proposed runtime boundary

```text
Deterministic tools
       ↓
canonical evidence
       ↓
known rule/signature
   ├─ matched → MACHINE result
   └─ unmatched/bounded
            ↓
     Local QA Coprocessor
            ↓
       policy gate
       ├─ safe/high confidence → result
       └─ unknown/complex → frontier agent
```

Model tidak boleh bypass deterministic failures atau mengubah FAIL menjadi PASS.

---

# Expected value

Target arah, **bukan SLA sebelum benchmark**:

```text
100 QA/build tasks
├─ majority deterministic
├─ subset local SLM triage
└─ minority frontier reasoning
```

Keberhasilan dinilai dari pengurangan total pekerjaan frontier tanpa menaikkan false-green, regression, atau debugging time.

---

# Non-goals

- membuat mini-Codex autonomous;
- mengganti CI/compiler/tests;
- memasukkan model weights ke Git;
- membuat model registry di repo ini;
- menambah local inference service sebelum ada measured need;
- melatih model hanya karena benchmark publik terlihat bagus;
- membuat ReleaseProof bergantung pada satu model vendor.

Model weights harus tetap di model store/cache/local disk, sementara repo menyimpan contract, benchmark schema, config, evidence, dan integration logic.

---

# Implementation order

## Phase A — foundation

- [x] selesaikan #3 sampai canonical evidence + agent packet
- [ ] tentukan shared SLM input/output schema
- [ ] tentukan escalation policy

## Phase B — benchmark

- [ ] bangun LocalQABench/ReleaseProofBench
- [ ] benchmark 3 kandidat tanpa fine-tuning
- [ ] catat latency/RAM/accuracy/failure modes

## Phase C — training experiment

- [ ] train `Qwen2.5-Coder-1.5B Base`
- [ ] compare against untrained baseline + Granite + Qwen3.5-0.8B
- [ ] reject training jika improvement tidak measurable

## Phase D — integration

- [ ] integrate sebagai optional `CHEAP_AGENT`
- [ ] default fallback ke deterministic/strong-agent path bila model unavailable
- [ ] retain structured evidence dan escalation reason

## Phase E — ReleaseProof consumer

- [ ] pilot terhadap #8 setelah ReleaseProof development gate terpenuhi
- [ ] gunakan build/release failures nyata sebagai verified evaluation set

---

# Acceptance criteria

RFC ini layak dipromosikan ke implementation ketika:

- [ ] canonical evidence + agent packet stabil;
- [ ] benchmark >=1,000 verified cases tersedia;
- [ ] candidate comparison selesai;
- [ ] Qwen2.5-Coder-1.5B atau kandidat pengganti menunjukkan measurable advantage;
- [ ] critical false-negative policy didefinisikan;
- [ ] SLM failure/unavailability tidak memblokir deterministic verification;
- [ ] no raw full-log requirement untuk normal failure;
- [ ] model weights tidak disimpan di Git;
- [ ] project-specific knowledge ownership mengikuti Project Integration Contract;
- [ ] frontier escalation selalu tersedia untuk unknown/complex cases.

---

## Decision

**Adopt as a planned extension of #3, but do not start model training/runtime integration before the P0 evidence + agent-packet foundation exists.**

Current preferred model: **Qwen2.5-Coder-1.5B Base**.

**Comment by Luciansvon** [comment]
Roadmap sync: ADR-0010 and merged PR #31 move Local QA Coprocessor/SLM to the final optional stage. No runtime or training starts before deterministic control-plane maturity, useful non-model adapters, Failure Memory baseline, a relevant consumer, and >=1,000 verified benchmark cases with protected held-out data. The model may never create PASS or authorize retries.
